WT-13186 Use file_cursors, not hs cursors when truncating the history store

ajmorton opened the pull request:
`__wt_session_range_truncate` expects `WT_CURSOR_BTREE`s to be provided when setting a truncation range, but `__wti_rts_history_btree_hs_truncate` provided `WT_CURSOR_HS` cursors instead.
This change updates `__wti_rts_history_btree_hs_truncate` to pass in the HS cursor's backing file cursors instead, which are WT_CURSOR_BTREE as expected.

Note this means we will store the the `file_cursor`s in `trunc_info` instead of the HS cursors. However all neccesary information is provided by the `file_cursor` so we'll note this is a special case when reading the cursor from trunc_info.

ajmorton commented:
I'm not overly familiar with the truncate code, so please suggest alternate solutions if you have them. 
@quchenhao I tried your suggestion to pass `file_cursor`s into `truncate()`, but this fails as functions like `__wt_curhs_range_truncate` expect `trunc_info->start` to contain HS cursors.

quchenhao commented on src/session/session_api.c:
```diff
@@ -1668,14 +1669,25 @@ __wt_session_range_truncate(
     trunc_info->orig_stop_key = orig_stop_key;
     trunc_info->uri = actual_uri;
 
+    if (start != NULL && strcmp(start->uri, WT_HS_URI) == 0)
```
Can we use WT_IS_HS macro?

ajmorton replied:
With the new approach we no longer need changes in `__wt_session_range_truncate`.
@pmacko86 I like your suggestion of `WT_CURSOR_IS_HS` macros. I'll add it to the scope of WT-13189

quchenhao commented on src/session/session_api.c:
```diff
@@ -1577,6 +1577,7 @@ int
 __wt_session_range_truncate(
   WT_SESSION_IMPL *session, const char *uri, WT_CURSOR *start, WT_CURSOR *stop)
 {
```
Can we ensure we pass the file cursor to this function? We should do the conversion in the upstream.

quchenhao replied:
rts_history.c
```
  if (!dryrun)
        WT_ERR(truncate_session->truncate(
          truncate_session, NULL, hs_cursor_start, hs_cursor_stop, NULL));
```

pmacko86 replied:
I'm not sure that we can assert that only {{WT_CURSOR_BTREE}} is passed to this function given that this is called pretty much directly from `WT_SESSION::truncate`, but if we are looking for a quick 8.0 fix, we could just change these lines in `rts_history.c` and add a FIXME to do a proper check for whether bounds are supported (as I explained in another comment).

pmacko86 commented on src/session/session_api.c:
```diff
@@ -1668,14 +1669,25 @@ __wt_session_range_truncate(
      * types at this abstraction level.
      */
-    if (CUR2BT(start) == NULL || CUR2BT(start)->type == BTREE_COL_FIX)
+    if (CUR2BT(start_file_cursor) == NULL || CUR2BT(start_file_cursor)->type == BTREE_COL_FIX)
```
It is a little confusing to introduce two new cursor variables just for this check. I'd say a better solution would be to incorporate checking for HS directly into these two `if` statements, such as something like this:

```
if (strcmp(start->uri, WT_HS_URI) == 0 || CUR2BT(start) == NULL || CUR2BT(start)->type == BTREE_COL_FIX)
```

(I don't think HS cursors support bounds—pls correct me if I'm wrong.) Adding a check here would make the condition more readable, and it would eliminate the need for the two new variables.

However, I still don't like this solution—a much better solution would check that `start` is a {{WT_CURSOR_BTREE}}. This function is called directly from `__session_truncate` (which is the public API `WT_SESSION::truncate`), so users can pass any kinds of cursors they want as long as they support `compare`. Ideally, this would become something like this:

```
if (!WT_CURSOR_IS_BTREE(start) || CUR2BT(start) == NULL || CUR2BT(start)->type == BTREE_COL_FIX)
```

...except that I don't know how to reliably check for the cursor type. Which I guess brings me to my other comments, where I wrote that I wish that `WT_CURSOR` would have a type field.

Alternatively, I wonder if there is a more generic way to check whether bounds are supported. I wish we could do something like `start->bound != __wt_cursor_config_notsup`, but that wouldn't work for complex types (column groups, tables with indexes). I wonder if it would make sense to have flags or some other ways to check that bounds are supported?

pmacko86 reviewed (commented):
Thank you for working on this! I provided some suggestions and alternatives in my comments for how to improve this solution.

When I added the check for `support_bounds`, I didn't realize that this code could be called with cursor types other than `WT_CURSOR_BTREE`. Sorry I missed it at that point 😞 I appreciate you fixing it!

pmacko86 commented:
I just noticed that [WT-13189](https://jira.mongodb.org/browse/WT-13189) was created, which is very much alongside of adding a type to `WT_CURSOR` that I was proposing. That would be the best fix for this issue. If you need something quick for 8.0, either fixing the call from `rts_history.c` or using `strcmp` directly in the `if` statement (plus adding the corresponding FIXME for when WT-13189 is ready) would be an okay quick fix if need be.

ajmorton commented:
@pmacko86 after discussion with @quchenhao and @kommiharibabu we've decided to update `__wti_rts_history_btree_hs_truncate` to pass in the `WT_CURSOR_BTREE` `file_cursor`s instead, similar to your suggestion [here](https://github.com/wiredtiger/wiredtiger/pull/10772#discussion_r1651844824). 
I've reverted the original changes and updated the description to reflect this.

kommiharibabu commented:
@ajmorton Thanks for the details and the fix. 

While going through the code one more time, A similar problem with the code where it converts the table cursor into the btree cursor in the `session_range_truncate` function. The test `test_truncate01` can generate this scenario. With the table cursors it is not causing any problem, but expecting it as a btree cursor is wrong and can cause problems.

pmacko86 reviewed (approved):
LGTM. Thank you.

ajmorton commented:
> @ajmorton Thanks for the details and the fix.
> 
> While going through the code one more time, A similar problem with the code where it converts the table cursor into the btree cursor in the `session_range_truncate` function. The test `test_truncate01` can generate this scenario. With the table cursors it is not causing any problem, but expecting it as a btree cursor is wrong and can cause problems.

Thanks @kommiharibabu. 
The reason this doesn't segfault is `offsetof(WT_CURSOR_BTREE, dhandle)` == `offsetof(WT_CURSOR_TABLE, table)`, and `WT_CURSOR_TABLE::table` is type `WT_TABLE` which begins with `WT_DATA_HANDLE iface;`. 
`CUR2BT` performs the wrong conversion, but we're still accessing the correct dhandle field in the end.

I can trigger a segfault with the following change to cursor.h
```diff
 struct __wt_cursor_table {
     WT_CURSOR iface;
 
+    int64_t buffer[50];
     WT_TABLE *table;
```
and adding `memset(ctable->buffer, -1, 50);` to `__wt_curtable_open`.

My callstack is
```
#0  0x0000fffff6be6720 in __wt_session_range_truncate (session=0x7d1ea0, uri=0x0, start=0x10402a0, stop=0xfd3ee0) at /home/ubuntu/src/wiredtiger/src/session/session_api.c:1676
#1  0x0000fffff6bf17f0 in __session_truncate (wt_session=0x7d1ea0, uri=0x0, start=0x0, stop=0xfd3ee0, config=0x0) at /home/ubuntu/src/wiredtiger/src/session/session_api.c:1878
#2  0x0000fffff6af8674 in _wrap_Session_truncate (self=0xfffff707d7b0, args=0xfffff61cee30) at lang/python/CMakeFiles/wiredtiger_python.dir/wiredtigerPYTHON_wrap.c:6894
```
so we're passing table cursors directly into the API as you mentioned yesterday. 

I think our best option is moving to the `WT_CURSOR::cursor_type` solution and adding special handling for each cursor type. I'll create a slack channel to discuss.

tod-johnson-mongodb commented:
The WT_CURSOR::cursor_type is great but we should go further. cursor_type checks are at execution time. We should make as many checks at compile time as possible. If a function is not part of the public API and it takes a WT_CURSOR * but that cursor must be only one type of cursor, that function should be changed to take the one type of cursor that it actually accepts. That will be detected at compile time.

tod-johnson-mongodb reviewed (approved):
LGTM - The changes solve the problems found so far.

ajmorton commented:
Thanks for the reviews all!
Based on slack discussions we'll drop this PR in favour of a different approach - either the `cursor->type` proposal above or WT-13206. We'll decide on this tomorrow and I'll update accordingly.

ajmorton commented:
Closing in favour of https://github.com/wiredtiger/wiredtiger/pull/10825
